### PR TITLE
Zoom Meetings (Independent Publisher) (Student)

### PR DESCRIPTION
**Pull request by A-Deekshith**
Microsoft Learn Student Ambassador Hackathon Submission Connector Hackfest

I would like to inform you that while Testing the connector it is showing Error: 403. But Everything is working fine. I had Tested it with Flow.

---
### When submitting a connector, please make sure that you follow the requirements below, otherwise your PR might be rejected. We want to make you have a well-built connector, a smooth certification experience, and your users are happy :) 

- [x] I attest that the connector works and I verified by deploying and testing all the operations. 
- [x] I attest that I have added detailed descriptions for all operations and parameters in the swagger file.
- [x] I attest that I have added response schemas to my actions, unless the response schema is dynamic. 
- [x] I validated the swagger file, `apiDefinition.swagger.json`, by running `paconn validate` command.
- [x] If this is a certified connector, I confirm that `apiProperties.json` has a valid brand color and doesn't use an invalid brand color, `#007ee5` or `#ffffff`. If this is an independent publisher connector, I confirm that I am not submitting a connector icon.

If you are an Independent Publisher, you must also attest to the following to ensure a smooth publishing process:
- [x] I have named this PR after the pattern of "Connector Name (Independent Publisher)" ex: HubSpot Marketing (Independent Publisher)
- [x] Within this PR markdown file, I have pasted screenshots that show: 3 unique operations (actions/triggers) working within a Flow. This can be in one flow or part of multiple flows. For each one of those flows, I have pasted in screenshots of the Flow succeeding. 
- [x] Within this PR markdown file, I have pasted in a screenshot from the Test operations section within the Custom Connector UI.
- [x] If the connector uses OAuth, I have provided detailed steps on how to create an app in the readme.md. 

Meetings into Excel with Calendar
Get Meetings
Meeting Details
Test Case

### CODE REVIEW

File: independent-publisher-connectors/Zoom Meetings/apiDefinition.swagger.json [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+  ],
+  "consumes": [],
+  "produces": [],
+  "paths": {
+    "/v2/users/me/meetings": {
+      "get": {
+        "responses": {
+          "default": {
+            "description": "default",
+            "schema": {}
```
Comment (sriyen-msft) [inline independent-publisher-connectors/Zoom Meetings/apiDefinition.swagger.json:21]: @A-Deekshith Empty response schema is not allowed anymore, which means no dynamic content will be shown in the output and makers have to use parse JSON to parse the response. Please provide valid response schema definition for all the operations in the swagger.

File: independent-publisher-connectors/Zoom Meetings/apiDefinition.swagger.json [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+                "settings": {
+                  "type": "object",
+                  "properties": {
+                    "host_video": {
+                      "type": "boolean",
+                      "description": "host_video"
+                    },
+                    "participant_video": {
+                      "type": "boolean",
+                      "description": "participant_video"
```
Comment (sriyen-msft) [inline independent-publisher-connectors/Zoom Meetings/apiDefinition.swagger.json:73]: @A-Deekshith All the description values are expected to be free from grammatical and spell errors. Please provide valid description values for all the properties.

File: independent-publisher-connectors/Zoom Meetings/apiProperties.json [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+            "refreshUrl": {
+              "value": "https://zoom.us/oauth/token"
+            }
+          }
+        }
+      }
+    },
+    "iconBrandColor": "#da3b01",
+    "capabilities": [],
+    "publisher": "Akuthota Deekshith"
```
Comment (sriyen-msft) [inline independent-publisher-connectors/Zoom Meetings/apiProperties.json:34]: @A-Deekshith Please add `stackOwner` property as well.

Docs - [Adding Metadata](https://docs.microsoft.com/connectors/custom-connectors/certification-submission#step-3a-publisher-and-stackowner-properties)

File: independent-publisher-connectors/Zoom Meetings/readme.md [hunk 4]
```diff
@@ -0,0 +1,26 @@
+# Zoom Connector
+Zoom is widely used for Virtual Meetings. Using this Connector you can Create Meetings, Get Meetings and Meeting Details (Individually). In some scenarios we want to perform certain actions Ex: Add Meetings into Excel and many more. In this process Zoom Connector will gets the work done.
+
+## Prerequisites
+You will need the following to proceed:
+•	A Microsoft Power Apps or Power Automate plan with custom connector feature
+•	A Zoom Account
```
Comment (sriyen-msft) [inline independent-publisher-connectors/Zoom Meetings/readme.md:7]: @A-Deekshith Please add a section called "Obtaining credentials and Setting up the application" and provide the instructions accordingly which would help us to setup an app to validate your connector.

**Review by sriyen-msft — CHANGES_REQUESTED** [review]

File: independent-publisher-connectors/Zoom Meetings/readme.md (hunk 4 above)
Comment (nk-gears) [reply independent-publisher-connectors/Zoom Meetings/readme.md:7]: @A-Deekshith, Please reach-out to me if you need any support on this connector. Wishing you all the best.

Comment (A-Deekshith) [reply independent-publisher-connectors/Zoom Meetings/readme.md:7]: @sriyen-msft @nk-gears Sir, I had done all the changes that you suggested. I request you to review the Files.

File: independent-publisher-connectors/Zoom Meetings/apiDefinition.swagger.json [hunk 5]
```diff
@@ -0,0 +1,430 @@
...
+        "summary": "Get Meetings",
+        "description": "Gives the Upcoming Meetings",
+        "operationId": "GetMetings",
+        "parameters": []
+      },
+      "post": {
+        "responses": {
+          "default": {
+            "description": "default",
+            "schema": {}
```
Comment (sriyen-msft) [inline independent-publisher-connectors/Zoom Meetings/apiDefinition.swagger.json:103]: @A-Deekshith Empty response schema is not allowed anymore, which means no dynamic content will be shown in the output and makers have to use parse JSON to parse the response. Please provide a valid response schema definition for all the operations in the swagger.

File: independent-publisher-connectors/Zoom Meetings/apiDefinition.swagger.json (hunk 2 above)
Comment (sriyen-msft) [reply independent-publisher-connectors/Zoom Meetings/apiDefinition.swagger.json:143]: @A-Deekshith Please address and fix this comment.

File: independent-publisher-connectors/Zoom Meetings/apiDefinition.swagger.json [hunk 6]
```diff
@@ -0,0 +1,430 @@
+{
+  "swagger": "2.0",
+  "info": {
+    "title": "Zoom Meetings",
+    "description": "This Connector will help to automate Zoom Meeting Operations",
+    "version": "1.0"
```
Comment (sriyen-msft) [inline independent-publisher-connectors/Zoom Meetings/apiDefinition.swagger.json:6]: @A-Deekshith The metadata is missing in your connector. Please add it as it is a requirement in the program.
Docs - [Adding Metadata](https://docs.microsoft.com/en-us/connectors/custom-connectors/certification-submission#step-3-add-metadata)

File: independent-publisher-connectors/Zoom Meetings/apiProperties.json [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+              "value": "https://zoom.us/oauth/token"
+            }
+          }
+        }
+      }
+    },
+    "iconBrandColor": "#da3b01",
+    "capabilities": [],
+    "publisher": "Akuthota Deekshith",
+    "stackOwner": "Zoom"
```
Comment (sriyen-msft) [inline independent-publisher-connectors/Zoom Meetings/apiProperties.json:35]: @A-Deekshith This service/API is owned by `"Zoom Video Communications, Inc."`, I am not sure why you have provided just "Zoom".

**Review by sriyen-msft — CHANGES_REQUESTED** [review]

File: independent-publisher-connectors/Zoom Meetings/apiProperties.json (hunk 7 above)
Comment (A-Deekshith) [reply independent-publisher-connectors/Zoom Meetings/apiProperties.json]: @sriyen-msft I had made the changes suggested by you.

**Comment by A-Deekshith** [comment]
@sriyen-msft Thanks for guiding me through this review process.

**Comment by sriyen-msft** [comment]
@A-Deekshith We have reviewed your connector and want to move forward with it, Please resolve all the comments.

File: independent-publisher-connectors/Zoom Meetings/readme.md [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+* Choose App as OAuth to create.
+* Enter Name and select Account-level app. Turn off the toggle.
+
+<b>Getting Credentials:</b>
+* In My Dashboard Go to Created Apps.
+* Select App Name
+* Here are your Credentials.
+* In "Redirect URL for OAuth" and "Add Allow List" add the Redirect URL of Custom Connector.
+* In "Add Allow List" insert https://us.flow.microsoft.com/en-us/
+
```
[the remark this answers is not in the document]
Comment (A-Deekshith) [reply independent-publisher-connectors/Zoom Meetings/readme.md:41]: @sriyen-msft I made all the necessary changes required. With the steps in the new Readme.md file error is fixed.

File: independent-publisher-connectors/Zoom Meetings/apiProperties.json [hunk 9]
```diff
@@ -0,0 +1,37 @@
+{
+  "properties": {
+    "connectionParameters": {
+      "token": {
+        "type": "oauthSetting",
+        "oAuthSettings": {
+          "identityProvider": "oauth2",
+          "clientId": "Zoom Client ID",
+          "scopes": [
+            "meeting:read:admin meeting:read"
```
[the remark this answers is not in the document]
Comment (A-Deekshith) [reply independent-publisher-connectors/Zoom Meetings/apiProperties.json]: @sriyen-msft Yes meeting:master. I updated that as well.

File: independent-publisher-connectors/Zoom Meetings/readme.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (A-Deekshith) [reply independent-publisher-connectors/Zoom Meetings/readme.md:41]: @sriyen-msft Do I have to make any other modifications?

[the remark this answers is not in the document]
Comment (sriyen-msft) [reply independent-publisher-connectors/Zoom Meetings/readme.md:41]: @A-Deekshith I have added below scopes to the OAUTH app and tried to test the connector locally. It is throwing an error saying "Invalid access token, does not contain scopes: [meeting:write, meeting:write:admin].". I have also tried to update the connector by adding these two write scopes to the connection settings, no luck. It seems you have also not submitted us a screenshot of a successful test of the "Create Meeting" operation. Please revisit the connector artifacts and try to reproduce the error and resolve it ASAP. 

image

image

**Review by sriyen-msft — CHANGES_REQUESTED** [review]

**Comment by A-Deekshith** [comment]
@sriyen-msft 

I had reverified again. 
In my case, it is working completely fine. These are screenshots regarding the Create Meeting in Flow as well as Connector. In Connector Testing there is a CORS Error but flow perfectly working
Create Meeting CORS
Flow Create Sucess

.

**Comment by sriyen-msft** [comment]
> @sriyen-msft
> 
> I had reverified again and created entirely new Connector and Zoom App using Readme.md Document. In my case, it is working completely fine. These are screenshots regarding the Create Meeting in Flow as well as Connector. In Connector Testing there is a CORS Error but Instant Cloud Flow is perfectly working.
> 
> Create Meeting CORS Flow Create Sucess
> 
> .

@A-Deekshith I have followed all the instructions you have mentioned in the readme.md to set up an OAuth app and created a custom connector to test it, It is throwing an "Invalid scope" error in the authentication. Please revisit the instructions you have provided and update them if required. We can not move forward until we see a successful test in a custom connector or test region.

image


image

**Comment by nk-gears** [comment]
@A-Deekshith , I am watching this thread for more than a month and i thought of helping you in the technical challenges. I can do a quick review and provide you suggestions related to oAuth Connections, Keys and Resources.  Please let me know and we can meet this Saturday.

**Comment by A-Deekshith** [comment]
> @A-Deekshith , I am watching this thread for more than a month and i thought of helping you in the technical challenges. I can do a quick review and provide you suggestions related to oAuth Connections, Keys and Resources. Please let me know and we can meet this Saturday.

Definitely, I'm ready to meet this Saturday. Thanks a lot for the support.

**Comment by A-Deekshith** [comment]
Based on the Discussion with @nk-gears. We concluded that scopes in the Connector Security Tab and Zoom App must Match and Few Other Steps.

**Comment by A-Deekshith** [comment]
@sriyen-msft I made few updates in Readme.md which fixed the invalid scope error.

**Comment by sriyen-msft** [comment]
> @sriyen-msft I made few updates in Readme.md which fixed the invalid scope error.

@A-Deekshith Sorry for the delay in response, We had a deployment freeze from last 2 weeks and I was OOF.

I have reviewed your updates in readme.md and performed a smoke test, it is working as expected. I will go ahead and queue up for this Wednesday's deployment schedule which is the last one of the year. I will keep you posted with the updates.

**Comment by A-Deekshith** [comment]
> > @sriyen-msft I made few updates in Readme.md which fixed the invalid scope error.
> 
> @A-Deekshith Sorry for the delay in response, We had a deployment freeze from last 2 weeks and I was OOF.
> 
> I have reviewed your updates in readme.md and performed a smoke test, it is working as expected. I will go ahead and queue up for this Wednesday's deployment schedule which is the last one of the year. I will keep you posted with the updates.

@sriyen-msft  That's Fine. Thank you for Reviewing the Connector.

**Comment by sriyen-msft** [comment]
@A-Deekshith Update - Your connector is now in production deployment and reached to [Test](https://docs.microsoft.com/en-us/connectors/custom-connectors/certification-submission#region-deployment) region, and it would be completed and available in all the production regions by Thursday 16th December 2021. Please let us know if you have any questions. Thanks

**Comment by sriyen-msft** [comment]
@A-Deekshith Apologies for the delay in response, there were a few deployment issues at our end which are resolved today.

Congratulations! The deployment of your connector and [documentation page](https://docs.microsoft.com/en-us/connectors/zoommeetingsip/) is now complete, We have a short [survey](https://forms.office.com/Pages/ResponsePage.aspx?id=v4j5cvGGr0GRqy180BHbR0BHMrjhL0hDmckROosW6AFURERIVUUzMk5XRzJGV1dJM1Q5RVZJUjFXSC4u) on the connector certification process that we would like for you to complete. Your feedback is appreciated and will help us improve the program, Thanks.

**Review by sriyen-msft — APPROVED** [review]

**Pull request merged**